### PR TITLE
Trovi 2.0: Implement native create/edit form

### PR DESCRIPTION
**Pull request by super-cooper**
The changes here are quite hefty, but there are still a couple things that don't work. 

* Ownership does not resolve properly. It used to work by looking up an artifact by path, but now that is not possible. I am unsure of the best way to get an artifact UUID to where ownership is determined. 
* I have implemented versioning, but have not tested it because ownership is not properly resolved. It was very simple to implement, similar to how the create calls are generated.
* Update has not been implemented yet. The form is set up such that it will be easy to edit/create a new version by reusing the same form, and it can be done at the same time. The only work is to format edits into JSON patches. Luckily, this is very simple to do with [this library](https://www.npmjs.com/package/rfc6902)

### CODE REVIEW

**Comment by super-cooper** [comment]
[Trovi 2.0: Implement native create/edit form](https://trello.com/c/mXzig1D6/2543-trovi-20-implement-native-create-edit-form)

File: jupyterlab_chameleon/artifact.py [hunk 1]
```diff
@@ -1,66 +1,257 @@
...
 import json
+import logging
 import os
+import requests
+import tarfile
 import tempfile
-import zipfile
-
-from keystoneauth1.exceptions.http import Unauthorized
+from dataclasses import asdict
```
Comment (diurnalist) [inline jupyterlab_chameleon/artifact.py:7]: OK, just ensure we depend on Python >= 3.7 explicitly for this

File: jupyterlab_chameleon/artifact.py [hunk 2]
```diff
@@ -1,66 +1,257 @@
...
+        self.created_at = created_at
+        self.slug = slug
+
+    @property
+    def contents_urn(self):
+        return f"urn:{self.contents_location}:{self.contents_id}"
+
+
+class Artifact:
+    """A shared experiment/research artifact that can be spawned in JupyterHub.
```
Comment (diurnalist) [inline jupyterlab_chameleon/artifact.py:66]: Docstring doesn't line up w/ implementation. TBH I don't think you really need documentation here, it's not a library, rather it is adapting a Trovi API object, if anything you could link to Trovi docs?

That said, you already have them, and they're useful for code inspection 👍🏻

File: jupyterlab_chameleon/artifact.py [hunk 3]
```diff
@@ -1,66 +1,257 @@
...
+        self.id = id
+        self.tags = tags or []
+        self.authors = authors or []
+        self.linked_projects = linked_projects or []
+        self.repro_enable_requests = repro_enable_requests
+        self.repro_access_hours = repro_access_hours
+        self.repro_requests = repro_requests
+        self.versions = versions or []
+
+        # Only the deposition information is required. Theoretically this can
```
Comment (diurnalist) [inline jupyterlab_chameleon/artifact.py:108]: Not sure how this comment relates to the exception, is this old text?

File: jupyterlab_chameleon/artifact.py [hunk 4]
```diff
@@ -1,66 +1,257 @@
...
     """
-    return call_jupyterhub_api('share/prepare_upload')
+    trovi_token = get_trovi_token()
+
+    response = {
+        "upload_endpoint": {
+            "url": contents_url(trovi_token),
+            "method": "POST",
+            "headers": {
+                "content-type": "application/json",
```
Comment (diurnalist) [inline jupyterlab_chameleon/artifact.py:132]: content-type is not JSON for an upload, won't it be a binary upload of a tarball?

File: src/artifact-sharing/widget.tsx [hunk 5]
```diff
@@ -164,10 +207,331 @@ class NewArtifactVersionSuccessText extends React.Component<ArtifactText.IProps>
...
+    )
+  }
+}
+
+class ArtifactEditForm extends React.Component<ArtifactForm.IProps> {
+  static readonly hidden = {display: 'none'};
+  static readonly block = {display: 'block'};
+
+  isCreateForm(): boolean {
+    return this.props.artifact.id === null || this.props.artifact.id === undefined;
```
Comment (diurnalist) [inline src/artifact-sharing/widget.tsx:395]:
```suggestion
    return !this.props.artifact.id;
```

**Review by diurnalist — CHANGES_REQUESTED** [review]
We can't drop the sqlite DB, it's the thing that binds the local server state (artifacts checked out to paths on the file system) to Trovi API state (artifact UUIDs). Other than that this is looking pretty good!

The sqlite DB is also the thing that ensures when you load the app, the filesystem icons turn green for folders that are linked to an artifact. It also ensures that when you refresh, the menu options are accurate (e.g., if I publish, then come back a day later, the options should understand it's "edit" or "new version".)

**Pull request merged**